morpheus/stages/input/rest_source_stage.py

```diff
@@ -0,0 +1,198 @@
+
+    def _parse_payload(self, payload: str) -> typing.Tuple[int, str]:
+        try:
+            # engine='cudf' is needed when lines=False to avoid using pandas
+            df = cudf.read_json(payload, lines=self._lines, engine='cudf')
+        except Exception as e:
+            err_msg = "Error occurred converting REST payload to Dataframe"
+            logger.error(f"{err_msg}: {e}")
+            return (400, MimeTypes.TEXT.value, err_msg, None)
+
+        try:
+            self._queue.put(df, block=True, timeout=self._queue_timeout)
+            return (201, MimeTypes.TEXT.value, "", None)
+        except (queue.Full, Closed) as e:
+            err_msg = "REST payload queue is "
+            if isinstance(e, queue.Full):
+                err_msg += "full"
+            else:
+                err_msg += "closed"
+            logger.error(err_msg)
+            return (503, MimeTypes.TEXT.value, err_msg, None)
+        except Exception as e:
+            err_msg = "Error occurred while pushing payload to queue"
+            logger.error(f"{err_msg}: {e}")
+            return (500, MimeTypes.TEXT.value, err_msg, None)
+
+    def _generate_frames(self) -> typing.Iterator[MessageMeta]:
+        from morpheus.common import FiberQueue
+        from morpheus.common import RestServer
+
+        self._queue = FiberQueue(self._max_queue_size)
+        rest_server = RestServer(parse_fn=self._parse_payload,
+                                 bind_address=self._bind_address,
+                                 port=self._port,
+                                 endpoint=self._endpoint,
+                                 method=self._method,
+                                 num_threads=self._num_server_threads,
+                                 max_payload_size=self._max_payload_size_bytes,
+                                 request_timeout=self._request_timeout_secs)
+        rest_server.start()
+
+        processing = True
+        while (processing):
+            # Read as many messages as we can from the queue if it's empty check to see if we should be shutting down
+            # It is important that any messages we received that are in the queue are processed before we shutdown since
+            # we already returned an OK response to the client.
+            df = None
+            try:
+                df = self._queue.get()
+            except queue.Empty:
+                if (not rest_server.is_running()):
+                    processing = False
+                else:
+                    logger.debug("Queue empty, sleeping ...")
+                    time.sleep(self._sleep_time)
+            except Closed:
+                logger.error("Queue closed unexpectedly, shutting down")
+                processing = False
+
+            if df is not None:
+                yield MessageMeta(df)
+
+        logger.debug("Stopping REST server ...")
+        rest_server.stop()
+        logger.debug("REST server stopped")
+        self._queue.close()
```

Comment:
Wrap this in a try...finally block to ensure the server gets closed.

Reply:
Or an even better option would be to have `RestServer` be a context manager

Reply:
Done, but now I'm wondering if we should add a context manager to `FiberQueue`